container_pipeline/workers/base.py

```diff
@@ -49,17 +50,26 @@ def run(self):
         while True:
             job_obj = self.queue.get()
             job = json.loads(job_obj.body)
-            debug_logs_file = os.path.join(
-                job['logs_dir'], settings.SERVICE_LOGFILE)
-            # Run dfh.clean() to clean log files if no error is encountered in
-            # post delivering build report mails to user
-            dfh = DynamicFileHandler(self.logger, debug_logs_file)
-            self.logger.info('Got job: {}'.format(job))
-            try:
-                self.handle_job(job)
-            except Exception as e:
-                self.logger.error(
-                    'Error in handling job: {}\nJob details: {}'.format(
-                        e, job), extra={'locals': locals()}, exc_info=True)
-            dfh.remove()
+
+            # Skip retrying a job if it's too early and push it back to queue.
+            # This will allow us to introduce some delays between job retries
+            if job.get('retry') is True and (
+                    time.time() - job.get('last_run_timestamp', 0) < job.get(
+                        'retry_delay', 0)):
+                self.queue.put(json.dumps(job), 'master_tube')
+            else:
+                debug_logs_file = os.path.join(
+                    job['logs_dir'], settings.SERVICE_LOGFILE)
+                # Run dfh.clean() to clean log files if no error is
+                # encountered in post delivering build report mails to user
+                dfh = DynamicFileHandler(self.logger, debug_logs_file)
+                self.logger.info('Got job: {}'.format(job))
+                try:
+                    self.handle_job(job)
+                except Exception as e:
+                    self.logger.error(
```

Comment:
We are not tracking any change in build state, it kind of complicates things. This approach is far simpler, you either process the job or not. The thing is that we don't log or process the job if it's being retried and the current time has not passed the delay window.